### PR TITLE
Query well known endpoints

### PR DESCRIPTION
**Pull request by karenzheung**
**Changes made this PR**

-  Capability querier queries the /.well-known/smart-configuration endpoint in addition to the /metadata 

- http status and response from well known endpoint added to Fhir_endpoints_info db 

- Small changes to message struct and FhirEndpointInfo struct 

**To test**

- `make run`
- populate db with endpoints `make populatedb`
- Send the endpoints to the queues: in `endpointmanager/cmd`, run `go run sendendpoints/main.go`
- Run the capability receiver: in `endpointmanager/cmd`, run `go run capabilityreceiver/main.go`
- Verify that the fhir_endpoints_info table in db has columns smart_http_response and smart_response
- Verify that some of the smart_http_response have 200 and that some of the smart_response have actual statements

Pull requests into this repository require the following checks to be completed by the submitter and reviewers.

**Submitter:**
- [x] Administrative tasks are complete.
  - This pull request describes why these changes were made.
  - The JIRA ticket links to this PR.
  - The JIRA ticket for this PR is: 
https://oncprojectracking.healthit.gov/support/browse/LANTERN-199
- [x] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.
- [x] Code review has been performed.
  - Code diff has been done and checked to ensure only expected code is being committed.
  - A linter for the language has been run.

**Primary Reviewer:**

- [x] Code works.
  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
- [x] Code review has been performed.
  - Code is maintainable, reusable, efficient, and correct.
  - Code accomplishes the tasks purpose.
  - Code follows style guidance appropriate for the language, including passing linter checks.
  - Code is well commented (which does not mean verbosely commented).
- [x] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.

**Secondary Reviewer:**

- [x] Code works.
  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
- [x] Code review has been performed.
  - Code accomplishes the tasks purpose.
- [x] Tests are complete.
  - Tests have been run locally and pass.

**Merger:**

Prior to merging, ensure that all checks have been completed. Delete the branch after the merge
is complete.

Create a new branch `<branch name>_gomodupdate`. Run `make update_mods` to update the `go.mod`
and `go.sum` files to point to master. Create a PR. Require one sanity check reviewer.

### CODE REVIEW

File: capabilityquerier/pkg/capabilityquerier/capabilityquerier.go [hunk 1]
```diff
@@ -47,11 +50,24 @@ func GetAndSendCapabilityStatement(
...
 	}
-
-	err = requestCapabilityStatement(ctx, fhirURL, client, &message)
+	// Query fhir endpoint
+	err = requestCapabilityStatement(ctx, metadataURL, "metadata", client, &message)
+	if err != nil {
+		message.Err = err.Error()
+	}
+	// Query well known endpoint
+	err = requestCapabilityStatement(ctx, wellKnownURL, "well-known", client, &message)
```
Comment (edeyoung) [inline capabilityquerier/pkg/capabilityquerier/capabilityquerier.go:70]: i recommend changing the name of this function as it now queries both the capability statement and the smart-on-fhir well-known endpoint.

File: capabilityquerier/pkg/capabilityquerier/capabilityquerier.go [hunk 2]
```diff
@@ -47,11 +50,24 @@ func GetAndSendCapabilityStatement(
 	queueName string) error {
 
 	var err error
+	metadataURL := &url.URL{}
+	*metadataURL = *fhirURL
+	metadataURL.Path = path.Join(metadataURL.Path, "metadata")
+
+	wellKnownURL := &url.URL{}
+	*wellKnownURL = *fhirURL
+	wellKnownURL.Path = path.Join(wellKnownURL.Path, "/.well-known/smart-configuration")
```
Comment (edeyoung) [inline capabilityquerier/pkg/capabilityquerier/capabilityquerier.go:59]: this will need to be updated based on matt's recent PR. shouldn't be too bad. I recommend creating a helper function for the well-known endpoint similar to what he has for metadata.

File: capabilityquerier/pkg/capabilityquerier/capabilityquerier.go [hunk 3]
```diff
@@ -71,13 +87,14 @@ func GetAndSendCapabilityStatement(
...
 // fills out message with http response code, tls version, capability statement, and supported mime types
-func requestCapabilityStatement(ctx context.Context, fhirURL *url.URL, client *http.Client, message *Message) error {
+func requestCapabilityStatement(ctx context.Context, fhirURL *url.URL, endptType string, client *http.Client, message *Message) error {
 	var err error
 	var httpResponseCode int
 	var supportsFHIR3MIMEType bool
 	var supportsFHIR2MIMEType bool
 	var tlsVersion string
 	var capResp []byte
+	var capabilityStatement interface{}
```
Comment (edeyoung) [inline capabilityquerier/pkg/capabilityquerier/capabilityquerier.go:97]: maybe call this `jsonResponse` or something more general

File: capabilityquerier/pkg/capabilityquerier/capabilityquerier.go [hunk 4]
```diff
@@ -113,12 +129,21 @@ func requestCapabilityStatement(ctx context.Context, fhirURL *url.URL, client *h
...
 		if err != nil {
 			return err
 		}
 	}
 
+	switch endptType {
+	case "metadata":
+		message.HTTPResponse = httpResponseCode
+		message.CapabilityStatement = capabilityStatement
+	case "well-known":
```
Comment (edeyoung) [inline capabilityquerier/pkg/capabilityquerier/capabilityquerier.go:142]: not a fan of "magic strings". can you make this an enum or similar at the top of the file?

File: capabilityquerier/pkg/capabilityquerier/capabilityquerier.go [hunk 5]
```diff
@@ -104,7 +121,6 @@ func requestCapabilityStatement(ctx context.Context, fhirURL *url.URL, client *h
 		}
 	}
 
-	message.HTTPResponse = httpResponseCode
 	message.TLSVersion = tlsVersion
```
Comment (edeyoung) [inline capabilityquerier/pkg/capabilityquerier/capabilityquerier.go:124]: does the TLSVersion, MimeType, etc. get overwritten depending on the order this is called?
I feel like we should either:
* write these once for the `metadata` call only
* capture TLS/MimeType/etc. explicitly for both the `metadata` and `well-known` calls.

Additionally, there's a fair amount of logic in here for calling the endpoint multiple times if a particular mime type doesn't work as expected. Again - we should either do that intentionally by capturing the different mime types for each endpoint type we query, or only do those duplicate calls once for the `metadata` endpoint.

I don't have a strong preference which we do... capturing for everything might be easiest because you just add filling out all of those fields into the switch statements you already have and then we have that data for both endpoints. However, we probably care the most about having that info for the metadata endpoints.

File: capabilityquerier/pkg/capabilityquerier/capabilityquerier_test.go [hunk 6]
```diff
@@ -47,7 +49,7 @@ func Test_GetAndSendCapabilityStatement(t *testing.T) {
 	// create the expected result
 	expectedCapStat, err := capabilityStatement()
 	th.Assert(t, err == nil, err)
-	expectedMimeType := []string{fhir2LessJSONMIMEType, fhir3PlusJSONMIMEType}
+	expectedMimeType := []string{fhir2LessJSONMIMEType, fhir3PlusJSONMIMEType, fhir2LessJSONMIMEType, fhir3PlusJSONMIMEType}
```
Comment (edeyoung) [inline capabilityquerier/pkg/capabilityquerier/capabilityquerier_test.go:52]: why are mime type entries being duplicated here?

Comment (karenzheung) [reply capabilityquerier/pkg/capabilityquerier/capabilityquerier_test.go]: The Mime types were added twice because the part of the code that gets this in requestCapabilityStatement function would be called twice. I changed it to only add the mime types for the metadata call only now

File: capabilityquerier/pkg/capabilityquerier/capabilityquerier.go (hunk 5 above)
Comment (karenzheung) [reply capabilityquerier/pkg/capabilityquerier/capabilityquerier.go]: Modified this so that the TLS and MimeType are only being captured for metadata endpoint

**Comment by karenzheung** [comment]
Fixed the wrong null entries in smart_response when it should have a statement

File: capabilityquerier/pkg/capabilityquerier/capabilityquerier.go [hunk 7]
```diff
@@ -98,35 +114,45 @@ func requestCapabilityStatement(ctx context.Context, fhirURL *url.URL, client *h
 		return err
```
Comment (bmath10) [inline capabilityquerier/pkg/capabilityquerier/capabilityquerier.go:115]: I think it might make sense for the wellknown endpoint call to use whatever mime type is currently saved in the message instead of just using the fhir3PlusJSONMIMEType, and if there isn't one saved to just use that one, since i would think it's more likely that whatever mime type worked for the metadata endpoint will also work for the wellknown one!

File: capabilityquerier/pkg/capabilityquerier/capabilityquerier_test.go [hunk 8]
```diff
@@ -58,14 +60,18 @@ func Test_GetAndSendCapabilityStatement(t *testing.T) {
...
 
 	// execute tested function
 	err = GetAndSendCapabilityStatement(ctx, fhirURL, &(tc.Client), &mq, &ch, queueName)
 	th.Assert(t, err == nil, err)
 	th.Assert(t, len(mq.(*mock.BasicMockMessageQueue).Queue) == 1, "expect one message on the queue")
 	message = <-mq.(*mock.BasicMockMessageQueue).Queue
+
+	var msgJSON map[string]interface{}
+	_ = json.Unmarshal(message, &msgJSON)
+	message, _ = json.Marshal(msgJSON["capabilityStatement"])
```
Comment (bmath10) [inline capabilityquerier/pkg/capabilityquerier/capabilityquerier_test.go:74]: I think instead of just testing the Capability Statement, the SMARTHTTPResponse and the SMARTResp should be added to the `expectedMsgStruct` and the whole message should be compared

File: endpointmanager/pkg/capabilityparser/factory.go [hunk 9]
```diff
@@ -26,6 +27,75 @@ type CapabilityStatement interface {
...
+	var err error
+	var respMsg map[string]interface{}
+
+	if len(respJSON) == 0 {
+		return nil, nil
+	}
+
+	err = json.Unmarshal(respJSON, &respMsg)
+	if err != nil {
+		return nil, errors.Wrap(err, "error unmarshalling JSON capability statement")
```
Comment (bmath10) [inline endpointmanager/pkg/capabilityparser/factory.go:67]: I think this error message is incorrect

File: endpointmanager/pkg/capabilityparser/factory.go [hunk 10]
```diff
@@ -26,6 +27,75 @@ type CapabilityStatement interface {
...
+
+func NewResponseBody(response map[string]interface{}) *ResponseBody {
+	return &ResponseBody{
+		Response: Response{
+			resp: response,
+		},
+	}
+}
+
+func NewSMARTRespFromInterface(response map[string]interface{}) (SMARTResponse, error) {
```
Comment (bmath10) [inline endpointmanager/pkg/capabilityparser/factory.go:50]: It looks like this function doesn't need to return an error

File: capabilityquerier/pkg/capabilityquerier/capabilityquerier.go [hunk 11]
```diff
@@ -53,13 +62,22 @@ func GetAndSendCapabilityStatement(
...
+	// Query fhir endpoint
+	err = requestCapabilityStatementAndSmartOnFhir(ctx, metadataURL, metadata, client, &message)
 	if err != nil {
 		log.Warnf("Got error:\n%s\n\nfrom URL: %s", err.Error(), fhirURL.String())
 		message.Err = err.Error()
 	}
 
+	wellKnownURL := endpointmanager.NormalizeWellKnownURL(fhirURL.String())
+	// Query well known endpoint
+	log.Info("WELL KNOWN URL IS " + wellKnownURL)
```
Comment (bmath10) [inline capabilityquerier/pkg/capabilityquerier/capabilityquerier.go:75]: Delete debugging statement

File: capabilityquerier/pkg/capabilityquerier/capabilityquerier.go (hunk 7 above)
Comment (karenzheung) [reply capabilityquerier/pkg/capabilityquerier/capabilityquerier.go:115]: In cases where both fhir3PlusJSONMIMEType and fhir2PlusJSONMIMEType are already stored in the message struct, should there a preference as to which is called? Or do both if the http status is not 200?

Comment (bmath10) [reply capabilityquerier/pkg/capabilityquerier/capabilityquerier.go:115]: for now (basing this off of lizzie's previous comment as well) i think i would just choose one and if it doesn't work then don't bother checking with the other

File: capabilityquerier/pkg/capabilityquerier/capabilityquerier_test.go (hunk 8 above)
Comment (karenzheung) [reply capabilityquerier/pkg/capabilityquerier/capabilityquerier_test.go:74]: In Test_GetAndSendCapabilityStatement it uses a specific test client from testClientWithContentType. It seems like it doesn't actually query it but gets the results from the testdata/.. directory. Should I do something similar and add a response statement from the well known endpoint since it doesn't actually get one?

Comment (bmath10) [reply capabilityquerier/pkg/capabilityquerier/capabilityquerier_test.go:74]: oh hmm okay, yeah that seems reasonable to me!

**Comment by bmath10** [comment]
It looks good to me, I'll approve once it's been rebased! Though I think the netstats querier PR is about to get merged in too so it's probably worthwhile to wait for that to happen before rebasing

**Comment by bmath10** [comment]
> It looks good to me, I'll approve once it's been rebased! Though I think the netstats querier PR is about to get merged in too so it's probably worthwhile to wait for that to happen before rebasing

It's been merged in if you want to rebase!
